lake/branch.go

```diff
@@ -123,6 +127,154 @@ func (b *Branch) Delete(ctx context.Context, ids []ksuid.KSUID, author, message
 	})
 }
 
+func (b *Branch) DeleteByPredicate(ctx context.Context, lake op.DataAdaptor, src string, author, message, meta string) (ksuid.KSUID, error) {
+	zctx := zed.NewContext()
+	appMeta, err := loadMeta(zctx, meta)
+	if err != nil {
+		return ksuid.Nil, err
+	}
+	val, op, err := compiler.ParseRangeExpr(zctx, src, b.pool.Layout)
+	if err != nil {
+		return ksuid.Nil, err
+	}
+	return b.commit(ctx, func(parent *branches.Config, retries int) (*commits.Object, error) {
+		base, err := b.pool.commits.Snapshot(ctx, parent.Commit)
+		if err != nil {
+			return nil, err
+		}
+		copies, deletes, err := b.getDbpObjects(ctx, lake, val, op, parent.Commit)
+		if err != nil {
+			return nil, err
+		}
+		count := len(copies) + len(deletes)
+		if count == 0 {
+			return nil, errors.New("nothing to delete")
+		}
+		// XXX Copied should be deleted if the commit fails.
```

Comment:
Do this in `dpbCopies`.
```suggestion
```

Reply:
There's a separate case where if the entire commit operation fails these should also be cleaned up, and this cannot happen in dpbCopies.